File: 147InsertionSortList/147InsertionSortList.cpp
#include <sstream>
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <unordered_map>
#include <algorithm>
#include <limits>

using namespace std;

/*
https://oj.leetcode.com/problems/insertion-sort-list/

Sort a linked list using insertion sort.
*/

// Definition for singly-linked list.
 struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(NULL) {}
 };
 
class Solution {
public:
    ListNode *insertionSortList(ListNode *head) {
   		if(head==nullptr) return nullptr;

   		ListNode *p=head;
   		ListNode *q=head;
   		ListNode *r=q->next;

   		while(r!=nullptr){
   			q=r;
   			r=r->next;

   			if(q->val<p->val){
   				p->next=r;

   				if(q->val<head->val){
   					q->next = head;
   					head = q;	
   				}else{
   					ListNode *o=nullptr;
   					ListNode *n=head;
   					while(n->val<=q->val){
   						o=n;
   						n=n->next;
   					}
   					o->next = q;
   					q->next = n;
   				}
   			}else{
   				p=q;
   			}
   		}     

   		return head;
    }
};

int main(){
	return 0;
}

File: solution/147InsertionSortList/147InsertionSortList.go
package main

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func insertionSortList(head *ListNode) *ListNode {
	dummy := ListNode{0, head}
	r := &dummy
	p := head
	for p != nil {
		q := p.Next

		l := &dummy
		o := dummy.Next
		for o != p && o.Val <= p.Val {
			l = l.Next
			o = o.Next
		}
		if o.Val > p.Val {
			l.Next = p
			p.Next = o
			r.Next = q
		} else {
			r = r.Next
		}
		p = q
	}

	return dummy.Next
}

func main() {
	//4,19,14,5,-3,1,8,5,11,15
}
